### PR TITLE
feat(Slider): Added new slider component

### LINKED ISSUE #5187: Add Slider component
Introduce a new Slider component.

Core issue and designs: https://github.com/patternfly/patternfly/issues/296 

There are two flavors of the slider: Continuous and Discrete. 

* For the continuous slider, the value returned should simply be based on the position between the start and end values. This can be fractional.

* For the discrete slider, when the user releases the mouse (or presses an arrow key to move the thumb) the thumb will always snap to the nearest tick mark.

### PR DESCRIPTION
**Pull request by tlabaj**
**What**: Closes #5187

TODO: integration tests

### CODE REVIEW

**Review by KKoukiou — COMMENTED** [review]
Thanks for working on this, we look forward to this feature in cockpit. 
I quickly tried it out, I have some pointers for improvement.

* Keyboard interactions don't work
* Clicking on random place in the slider should move the current position there

Tried it on firefox-83.0-13.fc33.x86_64.

**Review by mcarrano — CHANGES_REQUESTED** [review]
This is looking good @tlabaj . Just had a couple of issues/questions.

* For the continuous slider I notice that it never returns fractional values. Should we allow the consumer to set precision via a prop? There might be cases where you don't want the returned value always rounded to a whole number.
* For the "Thumb value input" example, I was able to get the thumb and value input to disappear from view by entering a value beyond the upper limit of the slider. Seems like there should be some value checking to make sure the slider value can never exceed the upper or lower limits.
* The keyboard interaction worked for me except on the last example of the slider with actions. @KKoukiou how were you expecting the interaction to work. We specified that the slider value should only change upon pressing Enter or tabbing out of the field. Not on every keystroke.
* Agree with @KKoukiou that you should also be able to update the value by clicking on the slider bar.

**Comment by KKoukiou** [comment]
> This is looking good @tlabaj . Just had a couple of issues/questions.
> 
>     * For the continuous slider I notice that it never returns fractional values. Should we allow the consumer to set precision via a prop? There might be cases where you don't want the returned value always rounded to a whole number.
> 
>     * For the "Thumb value input" example, I was able to get the thumb and value input to disappear from view by entering a value beyond the upper limit of the slider. Seems like there should be some value checking to make sure the slider value can never exceed the upper or lower limits.
> 
>     * The keyboard interaction worked for me except on the last example of the slider with actions. @KKoukiou how were you expecting the interaction to work. We specified that the slider value should only change upon pressing Enter or tabbing out of the field. Not on every keystroke.

I expect that clicking on the circle of the slider and then using my right and left keyboard keys, should move the slider value left and right accordingly. This does not happen for me at least. 
I just tried on all browsers i have available, firefox, chrome, edge and epiphany, its' the same everywhere.
Let me know if you still can't reproduce it.

> 
>     * Agree with @KKoukiou that you should also be able to update the value by clicking on the slider bar.

**Comment by tlabaj** [comment]
Thanks @KKoukiou and @mcarrano You are right, you have to tab into the thumb to be able to move it.  I will work on that update and clicking on the slider to update the value.

**Comment by tlabaj** [comment]
@mcarrano 

> This is looking good @tlabaj . Just had a couple of issues/questions.
> 
> * For the continuous slider I notice that it never returns fractional values. Should we allow the consumer to set precision via a prop? There might be cases where you don't want the returned value always rounded to a whole number.

It does return the fractional value.  The demo code takes the floor and displays that.  I can update that if you like.

> * The keyboard interaction worked for me except on the last example of the slider with actions. @KKoukiou how were you expecting the interaction to work. We specified that the slider value should only change upon pressing Enter or tabbing out of the field. Not on every keystroke.

For this one, are you talking about both the sliders in the action example or the very last one.  I could not reproduce an issue with the keyboard interaction other than what was mentioned about clicking on the slider and clicking the thumb to apply focus.

**Review by jessiehuff — COMMENTED** [review]
Really nice job with this so far, Titani!!! 🎉 ⭐ From my perspective, I think the keyboard interaction looks pretty good so far, and I think handling the aria position with the `aria-valuetext` was a decent way to do it. 

The only thing I've noticed that I thought was a little odd is that I can't seem to get the slider to 0 for a lot of the examples by keyboard, though it looks like this is true by mouse as well for some. Is it intentional to enforce a minimum value like that? Wondering if some of them may be a tad odd though. For example, if you're looking at the discrete example and the slider thumb starts at 2 and you drag it to 0, the slider value above it stays at 2 (I noticed this is true if it starts at any number and you drag it to 0, it will always stay at the previous number).

**Comment by mcarrano** [comment]
> For this one, are you talking about both the sliders in the action example or the very last one. I could not reproduce an issue with the keyboard interaction other than what was mentioned about clicking on the slider and clicking the thumb to apply focus.

Yes, I'm talking about the very last one. If I change the value in the number field and hit Enter or tab out of the field, nothing happens. I'm having the same problem with the second example under Value input.

**Review by jessiehuff — DISMISSED** [review]
LGTM! Looks like you fixed the thing I was seeing before. 🙂

File: packages/react-core/src/components/Slider/Slider.tsx [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+        type="number"
+        value={localInputValue}
+        aria-label={inputAriaLabel}
+        onKeyDown={handleKeyPressOnInput}
+        onChange={onChangeHandler}
+        onClick={onInputFocus}
+        onFocus={onInputFocus}
+        onBlur={onBlur}
+      />
+      {inputLabel && <span className={css(styles.inputGroupText, 'pf-m-plain')}>{inputLabel}</span>}
```
Comment (mcoker) [inline packages/react-core/src/components/Slider/Slider.tsx:253]: Technically an `<InputGroup>` is only needed here if this text exists, but it doesn't hurt to leave it the way it is.

File: packages/react-core/src/components/Slider/Slider.tsx [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+      thumbRef.current.style.setProperty('--pf-c-slider--value', `${newValue}%`);
+      setValue(newValue);
+      if (onValueChange) {
+        onValueChange(newValue);
+      }
+    }
+  };
+
+  const inputGroup = (
+    <div className={css(styles.inputGroup)}>
```
Comment (mcoker) [inline packages/react-core/src/components/Slider/Slider.tsx:240]: I'm curious why you don't use `<InputGroup>` here?

File: packages/react-core/src/components/Slider/Slider.tsx [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+      <input
```
Comment (mcoker) [inline packages/react-core/src/components/Slider/Slider.tsx:241]: Curious why you don't use `<TextInput>`?

File: packages/react-core/src/components/Slider/Slider.tsx (hunk 1 above)
Comment (mcoker) [reply packages/react-core/src/components/Slider/Slider.tsx:253]: Also curious why you don't use `<InputGroupText>` for this element? Looks like we would need an enhancement for `isPlain` or something to apply `.pf-m-plain`.

I can open a separate issue for that enhancement if we don't want to do it here. It was added as part of https://github.com/patternfly/patternfly/pull/3711/

**Review by mcoker — DISMISSED** [review]
LGTM! Left some questions, but they shouldn't hold up merging this PR.

**Review by mcarrano — APPROVED** [review]
Thanks for making those updates @tlabaj . Looks good to me!

File: packages/react-core/src/components/Slider/Slider.tsx (hunk 1 above)
Comment (tlabaj) [reply packages/react-core/src/components/Slider/Slider.tsx]: I have updated this to use iInputGroup, `InputGroupText` and `TextInput`.  It was an oversight.  

The `InputGroupText` component  still need to be updated for the plain variation.  @mcarrano do you want me to open a new issue to make that update?  The style is still applied inline here so it still displays correctly.

Comment (mcarrano) [reply packages/react-core/src/components/Slider/Slider.tsx]: Sure. Go ahead and open a core issue @tlabaj .

Comment (tlabaj) [reply packages/react-core/src/components/Slider/Slider.tsx]: @mcarrano it is a react issue.  Core already made the update.
here is the new issue
https://github.com/patternfly/patternfly-react/issues/5367

**Comment by kmcfaul** [comment]
Looks good, just two interaction questions.

In the `Thumb input value` example, you can move the slider but the text input doesn't update with the current slider value. The other slider examples do update so I'm not sure if this is intentional.

In the `Actions` example, you can "lock" the slider but only the text input is disabled, you can still drag the slider and change the value that way. Is the lock meant only for the text input?

**Comment by tlabaj** [comment]
@kmcfaul 
> Looks good, just two interaction questions.
> 
> In the `Thumb input value` example, you can move the slider but the text input doesn't update with the current slider value. The other slider examples do update so I'm not sure if this is intentional.

The input is working fine for me

> 
> In the `Actions` example, you can "lock" the slider but only the text input is disabled, you can still drag the slider and change the value that way. Is the lock meant only for the text input?
@mcarrano @mcoker there is no styling to disable a slider correct?

**Comment by mcarrano** [comment]
> In the Actions example, you can "lock" the slider but only the text input is disabled, you can still drag the slider and change the value that way. Is the lock meant only for the text input?
@mcarrano @mcoker there is no styling to disable a slider correct?

Right. There is no disabled styling for the Slider. We could add that but I'm hesitant as there is not a product use case for it. Can you just disable events on the slider for now @tlabaj to make this example work better?

**Comment by tlabaj** [comment]
> > In the Actions example, you can "lock" the slider but only the text input is disabled, you can still drag the slider and change the value that way. Is the lock meant only for the text input?
> > @mcarrano @mcoker there is no styling to disable a slider correct?
> 
> Right. There is no disabled styling for the Slider. We could add that but I'm hesitant as there is not a product use case for it. Can you just disable events on the slider for now @tlabaj to make this example work better?

@mcarrano I could.  However, I don't think that is a good idea.  I think that could be  confusing to the consumer since there is no visual indication that the slider is disabled.

**Comment by mcarrano** [comment]
@tlabaj @kmcfaul Do you think it's better to leave as is or to remove the example? Again, I see this as a "nice to have" feature that we don't have a hard requirement for right now.

**Review by kmcfaul — APPROVED** [review]
I think I'd be fine with it either way. Since this will be a beta component we can update the example later if we leave it in.

**Review by jessiehuff — COMMENTED** [review]
So I only see this on the actions example so it might just be that example's issue, but I noticed that when I try to navigate the actions example, it doesn't quite behave right (tring to use the right arrow key on the thumb doesn't work, only the left) and when I try to use the increase/plus button, it adds on a bizarre increment: 
Screen Shot 2021-01-25 at 1 21 09 PM

File: packages/react-core/src/components/Slider/Slider.tsx [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+}
+
+export interface SliderProps extends Omit<React.HTMLProps<HTMLDivElement>, 'onChange'> {
+  /** Additional classes added to the spinner. */
+  className?: string;
+  /** Current value  */
+  currentValue?: number;
+  /** Flag indicating if the slider is is discrete */
+  isDiscrete?: boolean;
+  /** Array of slider step objects (value and label of each step) for the slider. If this is provided, the numSteps prop will be ignored.*/
```
Comment (nicolethoen) [inline packages/react-core/src/components/Slider/Slider.tsx:25]: is there still a `numSteps` prop? I dont see it.

**Comment by nicolethoen** [comment]
we should also open an issue to make the sliders work on mobile.

**Review by jessiehuff — APPROVED** [review]
LGTM on the accessibility side! 🙂

**Comment by tlabaj** [comment]
@nicolethoen there is already an issue open for that.

File: packages/react-core/src/components/Slider/Slider.tsx (hunk 4 above)
Comment (tlabaj) [reply packages/react-core/src/components/Slider/Slider.tsx:25]: nope. that was removed

**Review by nicolethoen — APPROVED** [review]

**Review by kmcfaul — APPROVED** [review]

**Pull request merged**